src/Brahma.FSharp.OpenCL.Core/Core.fs

```diff
@@ -18,89 +18,174 @@ module Brahma.FSharp.OpenCL.Core
         let args = ref [||]
         let run = ref Unchecked.defaultof<Commands.Run<'TRange>>
+
         let getStarterFuncton qExpr =
-            let rec go expr vars =
-                match expr with
-                | Patterns.Lambda (v, body) ->
-                    Expr.Lambda(v, go body (v::vars))
-                | e ->
-                    let arr =
-                        let c = Expr.NewArray(typeof<obj>,vars |> List.rev
-                            |> List.map
-                                 (fun v -> Expr.Coerce (Expr.Var(v), typeof<obj>)))
-                        <@@
-                            let x = %%c |> List.ofArray
-                            rng := (box x.Head) :?> 'TRange
-                            args := x.Tail |> Array.ofList
-                            let brahmsRunCls = new Brahma.OpenCL.Commands.Run<_>(kernel,!rng)
-                            !args |> Array.iteri (fun i x -> brahmsRunCls.SetupArgument(1,i,x))
-                            run := kernel.Run(!rng, !args)
-                        @@>
-                    arr
-            let res = <@ %%(go qExpr []):'TRange ->'a @>.Compile()
-
-            res
-
-        if _outCode.IsSome then (_outCode.Value) := (kernel :> ICLKernel).Source.ToString()
-
-        kernel
-        , getStarterFuncton query
-        , fun () -> !run
+            match qExpr with
+            | DerivedPatterns.Lambdas (lambdaArgs, _) ->
+                let flattenArgs = List.collect id lambdaArgs
+
+                let firstMutexIdx =
+                    flattenArgs
+                    |> List.tryFindIndex (fun v -> v.Name.EndsWith "Mutex")
+                    |> Option.defaultValue flattenArgs.Length
+
+                let argsWithoutMutexes = flattenArgs.[0 .. firstMutexIdx - 1]
+
+                /// For each atomic variable throws exception if variable's type is not array,
+                /// otherwise returns length of array
+                let mutexLengths =
+                    let atomicVars =
+                        List.init<Var> (flattenArgs.Length - firstMutexIdx) <| fun i ->
+                            let mutexVar = flattenArgs.[firstMutexIdx + i]
+                            argsWithoutMutexes |> List.find (fun v -> mutexVar.Name.Contains v.Name)
+
+                    Expr.NewArray(
+                        typeof<int>,
+
+                        atomicVars
+                        |> List.map
+                            (fun var ->
+                                match var with
+                                | var when var.Type.IsArray ->
+                                    Expr.PropertyGet(
+                                        Expr.Var var,
+                                        typeof<int[]>.GetProperty("Length")
+                                    )
+                                | _ -> failwith "Non-array variables as global mutex parameters is not supported. Wrap it into array instead."
```

Comment:
Could this problem be not related to the translator? Or, if this exception is thrown, then we have a bug in the translator.